crates/wasmtime/src/runtime/types.rs

```diff
@@ -207,35 +214,54 @@ impl FuncType {
     /// The function descriptor returned will represent a function which takes
     /// `params` as arguments and returns `results` when it is finished.
     pub fn new(
+        engine: &Engine,
         params: impl IntoIterator<Item = ValType>,
         results: impl IntoIterator<Item = ValType>,
     ) -> FuncType {
-        FuncType {
-            sig: WasmFuncType::new(
+        Self::from_wasm_func_type(
+            engine,
+            &WasmFuncType::new(
                 params.into_iter().map(|t| t.to_wasm_type()).collect(),
                 results.into_iter().map(|t| t.to_wasm_type()).collect(),
             ),
-        }
+        )
     }
 
     /// Returns the list of parameter types for this function.
     #[inline]
     pub fn params(&self) -> impl ExactSizeIterator<Item = ValType> + '_ {
-        self.sig.params().iter().map(ValType::from_wasm_type)
+        self.ty.params().iter().map(ValType::from_wasm_type)
     }
 
     /// Returns the list of result types for this function.
     #[inline]
     pub fn results(&self) -> impl ExactSizeIterator<Item = ValType> + '_ {
-        self.sig.returns().iter().map(ValType::from_wasm_type)
+        self.ty.returns().iter().map(ValType::from_wasm_type)
+    }
+
+    pub(crate) fn type_index(&self) -> VMSharedTypeIndex {
+        self.ty.index()
     }
 
     pub(crate) fn as_wasm_func_type(&self) -> &WasmFuncType {
-        &self.sig
+        &self.ty
+    }
+
+    pub(crate) fn into_registered_type(self) -> RegisteredType {
+        self.ty
     }
 
-    pub(crate) fn from_wasm_func_type(sig: WasmFuncType) -> FuncType {
-        Self { sig }
+    pub(crate) fn from_wasm_func_type(engine: &Engine, ty: &WasmFuncType) -> FuncType {
+        let ty = engine.signatures().register(ty);
+        Self { ty }
+    }
```

Comment:
This seems like a method we should be ideally removing through this refactoring (or enabling removing).

The callers I see are:

* `ExternType::from_wasmtime` - this seems like it should not be necessary since the signature is guaranteed to aleady be registered through the module already so with a bit more plumbing we should be able to convert to a shared index in theory.
* `ComponentItem::from` - this is I think the same as the above just with components instead of modules
* `FuncType::new` - this seems justified, but in the sense that this method would get inlined into `FuncType::new`

It's ok to defer this to later, but wanted to note this down.

Reply:
The first two should still re-register the type (ie increment its ref count), even though it should already be registered, because we wouldn't want to end up in the situation where

* we load a module, registering its types
* get a registered type from the module
* drop the module, unregistering its types
* and now have a dangling function type

We could have a helper to get an already-registered type and increment its ref count, but I'm not sure how that would be any more efficient than what exists now. I guess it could at minimum assert that the type is already registered, but this isn't really giving us much I don't think.

Or maybe I am completely missing your point?